core/lib/zksync_core/src/api_server/execution_sandbox/execute.rs

```diff
@@ -108,40 +109,14 @@ pub(crate) async fn execute_tx_eth_call(
     vm_result
 }
 
-#[tracing::instrument(skip_all)]
-pub(crate) async fn execute_tx_with_pending_state(
-    vm_permit: VmPermit,
-    mut shared_args: TxSharedArgs,
-    execution_args: TxExecutionArgs,
-    connection_pool: ConnectionPool,
-    tx: Transaction,
-) -> (VmExecutionResultAndLogs, TransactionExecutionMetrics) {
-    let mut connection = connection_pool.access_storage_tagged("api").await.unwrap();
-    let block_args = BlockArgs::pending(&mut connection).await;
-    drop(connection);
-    // In order for execution to pass smoothlessly, we need to ensure that block's required gasPerPubdata will be
-    // <= to the one in the transaction itself.
-    shared_args.adjust_l1_gas_price(tx.gas_per_pubdata_byte_limit());
-
-    execute_tx_in_sandbox(
-        vm_permit,
-        shared_args,
-        execution_args,
-        connection_pool,
-        tx,
-        block_args,
-        vec![],
-    )
-    .await
-}
-
 /// This method assumes that (block with number `resolved_block_number` is present in DB)
 /// or (`block_id` is `pending` and block with number `resolved_block_number - 1` is present in DB)
 #[allow(clippy::too_many_arguments)]
 #[tracing::instrument(skip_all)]
-async fn execute_tx_in_sandbox(
+pub(crate) async fn execute_tx_in_sandbox(
     vm_permit: VmPermit,
     shared_args: TxSharedArgs,
+    adjust_pubdata_price: bool,
```

Comment:
please add a comment on this field. When do we need to adjust the pubdata price? When not?

Reply:
thinking more about this - I think it's a little bit of a code smell that we need to do something like this. So we pass an instance of `TxSharedArgs` where fee-related fields are _sometimes_ incomplete. Fixing this might require larger refactoring. 

Another code smell here (not introdeuced by this PR) is that we need to pass this `apply` lambda function. Did anyone already thing how this can be improved? cc @popzxc 

This is of course not for this PR

Reply:
Re `TxSharedArgs` / `apply`: probably we need a enum here.
Re `apply` lambda: I think it's caused by the fact that VM initialization is complex and very customizable. Probably it's possible to make it nicer, but we've already had 3 or 4 rounds of refactoring this module, and `apply` closure outlived all of them, so I'd say that likely it won't be easy to get rid of it.